lib/union_of.rb

```diff
@@ -0,0 +1,422 @@
+    def delete_records(...) = raise ReadonlyAssociationError.new(owner, reflection)
+    def concat_records(...) = raise ReadonlyAssociationError.new(owner, reflection)
+    def build_record(...)   = raise ReadonlyAssociationError.new(owner, reflection)
+  end
+
+  class Association < ReadonlyAssociation
+    def association_scope
+      return if
+        klass.nil?
+
+      @association_scope ||= Scope.create.scope(self)
+    end
+  end
+
+  module Preloader
+    class LoaderRecords < ActiveRecord::Associations::Preloader::Association::LoaderRecords; end
+    class LoaderQuery < ActiveRecord::Associations::Preloader::Association::LoaderQuery
+      attr_reader :reflection,
+                  :model
+
+      def initialize(scope, reflection, model)
+        @scope      = scope
+        @reflection = reflection
+        @model      = model
+      end
+
+      def records_for(loaders)
+        LoaderRecords.new(loaders, self).records
+      end
+
+      def load_records_for_keys(keys, &block)
+        foreign_reflection = reflection.inverse_of
+
+        raise MissingInverseForEagerLoadedUnionError.new(model, reflection) if
+          foreign_reflection.nil?
+
+        klass         = reflection.klass
+        table         = klass.arel_table
+        foreign_klass = foreign_reflection.klass
+        foreign_table = foreign_klass.arel_table
+        primary_key   = reflection.active_record_primary_key
+        foreign_key   = reflection.join_foreign_key
+        union_sources = foreign_reflection.union_sources
+
+        unions = union_sources.reduce(nil) do |left, union_source|
+          union_source_reflection  = klass.reflect_on_association(union_source)
+          union_source_foreign_key = union_source_reflection.join_foreign_key
+
+          right = case
+                  when union_source_reflection.through_reflection?
+                    through_reflection  = union_source_reflection.through_reflection
+                    through_foreign_key = through_reflection.join_foreign_key
+                    through_klass       = through_reflection.klass
+                    through_table       = through_klass.arel_table
+
+                    table.project(table[primary_key])
+                         .from(table)
+                         .join(through_table, Arel::Nodes::InnerJoin)
+                         .on(
+                           table[foreign_key].eq(through_table[through_foreign_key]),
+                         )
+                         .where(
+                           through_table[union_source_foreign_key].in(keys),
+                         )
+                  when union_source_reflection.belongs_to?
+                    table.project(table[primary_key])
+                         .from(table)
+                         .join(foreign_table, Arel::Nodes::InnerJoin)
+                         .on(
+                           table[union_source_foreign_key].eq(foreign_table[primary_key]),
```

Comment:
This is using the wrong primary key. It should be `foreign_reflection.active_record_primary_key`.